### PR TITLE
Added a collection of Ulid string related methods

### PR DESCRIPTION
**Pull request by keithy**
For review and discussion.

Ulid::fromString( $id )
Ulid::sanitizeString( $id )
Ulid::fromString( Ulid::sanitizeString( $id) );
Ulid::isStringAnUlid( $id );

### CODE REVIEW

**Comment by keithy** [comment]
Added a MalformdUlid exception

**Comment by tuupola** [comment]
Hi, currently this PR introduces too many things. There should be only one feature or change per pull request. I do like the `fromString()` method. However it should be in separate factory class similar as [KsuidFactory](https://github.com/tuupola/ksuid/blob/2.x/src/KsuidFactory.php). Will add some more comments into the source.

File: src/Ulid.php [hunk 1]
```diff
@@ -25,8 +25,19 @@ class Ulid
...
-    public function __construct($timestamp = null, $payload = null)
+    // Given that the majority of the time an Ulid is used as a string,
+    // the current internal representation is arguably the least efficient.
+    // This approach to constuction allows future improvements to be made without
+    // changing the interface.
+
+    public function __construct($timestamp = null, $payload = null, $alreadyEncoded = false)
     {
+
+        if ($alreadyEncoded) {
```
Comment (tuupola) [inline src/Ulid.php]: This adds confusion without any improvements and should be removed.

File: src/Ulid.php [hunk 2]
```diff
@@ -87,4 +92,58 @@ public function __toString()
     {
         return $this->string();
     }
+
+    //"0123456789ABCDEFGHJKMNPQRSTVWXYZ"
+    protected static function crockford()
```
Comment (tuupola) [inline src/Ulid.php]: I am not fan of static methods. To avoid instantiating the base32 encoder twice it could be instantiated in the constructor instead.

```php
private $base32;
...
$this->base32 = new \Tuupola\Base32([
    "characters" => \Tuupola\Base32::CROCKFORD,
    "padding" => false
]);
```

And then it can be used later as:

```php
$encoded = $this->base32->encode($this->timestamp);
```

File: src/Ulid.php [hunk 3]
```diff
@@ -87,4 +92,58 @@ public function __toString()
...
+
+        return new static($timestamp, $payload, true);
+    }
+
+    public static function isStringAnUlid($value)
+    {
+        return (1 === preg_match("/^[0-7][0-9A-HJKMNP-TV-Z]{25}$/", $value));
+    }
+
+    public static function mustBeAnUlid($str, $strict = true)
```
Comment (tuupola) [inline src/Ulid.php]: This is not needed.

File: src/Ulid.php [hunk 4]
```diff
@@ -87,4 +92,58 @@ public function __toString()
...
+        if (!isset($base32)) {
+            $base32 = new \Tuupola\Base32([
+                "characters" => \Tuupola\Base32::CROCKFORD,
+                "padding" => false
+            ]);
+        }
+        return $base32;
+    }
+
+    public static function sanitizeString($value)
```
Comment (tuupola) [inline src/Ulid.php]: This should be a separate PR. As a sidenote it is not 100% clear if ULID spec requires this even though is is part of Crockford standard.

File: src/Ulid.php [hunk 5]
```diff
@@ -87,4 +92,58 @@ public function __toString()
...
+    public static function fromString($value)
+    {
+
+        $timestamp = substr($value, 0, 10);
+        $payload = substr($value, 10);
+
+        return new static($timestamp, $payload, true);
+    }
+
+    public static function isStringAnUlid($value)
```
Comment (tuupola) [inline src/Ulid.php]: Not sure if this is needed. Instead `UlidFactory::fromString()` should throw if given an invalid string.
